storage/store.go

```diff
@@ -1533,19 +1545,107 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
 			GIDMap:         copySlicePreferringNil(gidMap),
 		}
 	}
-	return rlstore.create(id, parentLayer, names, mountLabel, nil, &options, writeable, diff, slo)
+	return &options, parentLayer, nil
 }
 
 func (s *store) PutLayer(id, parent string, names []string, mountLabel string, writeable bool, lOptions *LayerOptions, diff io.Reader) (*Layer, int64, error) {
 	rlstore, rlstores, err := s.bothLayerStoreKinds()
 	if err != nil {
 		return nil, -1, err
 	}
+
+	var (
+		slo         *stagedLayerOptions
+		options     *LayerOptions
+		parentLayer *Layer
+	)
+
+	if diff != nil {
+		m := rlstore.newMaybeStagedLayerExtraction(diff)
+		defer func() {
+			if err := m.cleanup(); err != nil {
+				logrus.Errorf("Error cleaning up temporary directories: %v", err)
+			}
+		}()
+		// driver can do unlocked staging so do that without holding the layer lock
+		if m.staging != nil {
+			// func so we have a scope for defer, we don't want to hold the lock for stageWithUnlockedStore()
+			layer, err := func() (*Layer, error) {
+				if err := rlstore.startWriting(); err != nil {
+					return nil, err
+				}
+				defer rlstore.stopWriting()
+
+				if layer, err := rlstore.checkIdOrNameConfict(id, names); err != nil {
+					return layer, err
+				}
+
+				options, parentLayer, err = populateLayerOptions(s, rlstore, rlstores, parent, lOptions)
+				return nil, err
+			}()
+			if err != nil {
+				return layer, -1, err
+			}
+
+			// make sure to use the resolved full ID if there is a parent
+			if parentLayer != nil {
+				parent = parentLayer.ID
+			}
+
+			err = rlstore.stageWithUnlockedStore(m, parent, options)
+			if err != nil {
+				return nil, -1, err
+			}
+		}
+
+		slo = &stagedLayerOptions{
+			stagedLayerExtraction: m,
+		}
+	}
+
 	if err := rlstore.startWriting(); err != nil {
 		return nil, -1, err
 	}
 	defer rlstore.stopWriting()
-	return s.putLayer(rlstore, rlstores, id, parent, names, mountLabel, writeable, lOptions, diff, nil)
+
+	if parentLayer == nil {
+		// options where not populated yet
+		options, parentLayer, err = populateLayerOptions(s, rlstore, rlstores, parent, lOptions)
+		if err != nil {
+			return nil, -1, err
+		}
+	} else {
+		// We used the staged extraction without holding the lock.
+		// Check again that the parent layer is still valid and exists.
+		freshLayer, err := getParentLayer(rlstore, rlstores, parentLayer.ID)
+		if err != nil {
+			return nil, -1, err
+		}
+		if !compareIdMappings(freshLayer.UIDMap, parentLayer.UIDMap) || !compareIdMappings(freshLayer.GIDMap, parentLayer.GIDMap) {
+			// Fatal problem. Mappings changed so the parent must be considered different now.
+			// Since we consumed the diff there is no we to recover, return error to caller. The caller would need to retry.
+			// How likely is that and would need to return a special error so c/image could do the retries?
```

Comment:
(This should currently not happen for c/image-created layers, because those have content-based IDs. It could happen only with a non-c/image writer that does something different. But I don’t know whether we will continue to have deterministic IDs with the sha512 work, I suspect we might not.

Either way, c/image pulls are currently not resilient to concurrent layer removals, so a hard failure here is ~fine. We’d need some kind of “lease a layer” feature that allows the pull to hold a found parent layer while downloading a child, and that does not exist.)